### ISSUE TITLE
Vector.Dot<float> is slower than trivial product and sum implementation (with repro)

### ISSUE THREAD

**Issue by goldshtn** [dotnet/coreclr#1414]
On my machine (Intel Core i7-940, Nehalem) the following implementations of float vector dot-product produce the following timings:

```
VectorDP 20.36 ms
CustomDP 14.89 ms
```

This is a pretty serious difference. Here's a portion of the relevant disassembly:

```
C:\Temp\DPRepro\Program.cs @ 42:
00007ff9`20f51386 448d5003        lea     r10d,[rax+3]
00007ff9`20f5138a 453bd0          cmp     r10d,r8d
00007ff9`20f5138d 733f            jae     00007ff9`20f513ce
00007ff9`20f5138f 0f104c8110      movups  xmm1,xmmword ptr [rcx+rax*4+10h]

C:\Temp\DPRepro\Program.cs @ 43:
00007ff9`20f51394 453bd1          cmp     r10d,r9d
00007ff9`20f51397 7335            jae     00007ff9`20f513ce
00007ff9`20f51399 0f10548210      movups  xmm2,xmmword ptr [rdx+rax*4+10h]

C:\Temp\DPRepro\Program.cs @ 44:
00007ff9`20f5139e 0f59ca          mulps   xmm1,xmm2
00007ff9`20f513a1 0f28d9          movaps  xmm3,xmm1
00007ff9`20f513a4 0fc6dbb1        shufps  xmm3,xmm3,0B1h
00007ff9`20f513a8 0f58cb          addps   xmm1,xmm3
00007ff9`20f513ab 0f28d9          movaps  xmm3,xmm1
00007ff9`20f513ae 0fc6db1b        shufps  xmm3,xmm3,1Bh
00007ff9`20f513b2 0f58cb          addps   xmm1,xmm3
00007ff9`20f513b5 f30f58c1        addss   xmm0,xmm1

C:\Temp\DPRepro\Program.cs @ 40:
00007ff9`20f513b9 83c004          add     eax,4
00007ff9`20f513bc 443bc0          cmp     r8d,eax
00007ff9`20f513bf 7fc5            jg      00007ff9`20f51386
```

Note that the bounds checks are not elided. I am also not sure about the instruction selection -- two pairs of shufps + addps to calculate the dot product.

The complete repro follows:

```C#
class Program
{
    static Random _rand = new Random();

    static float[] GetRandomVector(int size)
    {
        float[] result = new float[size];
        for (int i = 0; i < result.Length; ++i)
            result[i] = (float)_rand.NextDouble();
        return result;
    }

    static void Measure(Action<float[], float[]> action, float[] a, float[] b)
    {
        action(a, b); // warm up
        Stopwatch sw = Stopwatch.StartNew();
        for (int i = 0; i < 10; ++i)
        {
            action(a, b);
        }
        Console.WriteLine("{0} {1:N2} ms", action.Method.Name, sw.Elapsed.TotalMilliseconds);
    }

    static volatile float _dummy;

    static void VectorDP(float[] a, float[] b)
    {
        float dp = 0.0f;
        int vecSize = Vector<float>.Count;
        for (int i = 0; i < a.Length; i += vecSize)
        {
            Vector<float> va = new Vector<float>(a, i);
            Vector<float> vb = new Vector<float>(b, i);
            dp += Vector.Dot(va, vb);
        }
        _dummy = dp;
    }

    static void CustomDP(float[] a, float[] b)
    {
        Vector<float> vdp = new Vector<float>(0.0f);
        int vecSize = Vector<float>.Count;
        for (int i = 0; i < a.Length; i += vecSize*2)
        {
            Vector<float> va1 = new Vector<float>(a, i);
            Vector<float> vb1 = new Vector<float>(b, i);
            Vector<float> vprod1 = va1 * vb1;
            Vector<float> va2 = new Vector<float>(a, i);
            Vector<float> vb2 = new Vector<float>(b, i);
            vdp += vprod1;
            Vector<float> vprod2 = va2 * vb2;
            vdp += vprod2;
        }
        float dp = 0.0f;
        for (int i = 0; i < vecSize; ++i)
            dp += vdp[i];
        _dummy = dp;
    }

    static void Main(string[] args)
    {
        float[] a = GetRandomVector(1024 * 1024);
        float[] b = GetRandomVector(1024 * 1024);

        VectorDP(a, b); Console.WriteLine(_dummy);
        CustomDP(a, b); Console.WriteLine(_dummy);

        Measure(VectorDP, a, b);
        Measure(CustomDP, a, b);
    }
}
```

**Comment by mellinoe** [comment]
Thanks for the in-depth code. Small question/clarification: should the second pair of vectors be created with an offset in your custom method?

````CSharp
Vector<float> va2 = new Vector<float>(a, i + vecSize);
Vector<float> vb2 = new Vector<float>(b, i + vecSize);
````

Is there any significant difference if the custom implementation is further unrolled?

@CarolEidt could respond more to the generated assembly.

**Comment by mikedn** [comment]
I fail to see what the issue is, you're comparing apples and oranges. The custom implementation is what an autovectorizer may produce, it's not something that can be expected from an intrinsic based implementation.

**Comment by goldshtn** [comment]
@mellinoe Yes, of course it should be `i + vecSize`. Fixing the repro code above. The timings are still in effect, pretty much the same perf difference of 20-40% across a few runs.

Unrolling further actually seems to hurt performance.

**Comment by goldshtn** [comment]
@mikedn Well, `Vector.Dot` could use the trivial implementation with `mulps` and `addps` instead of emitting `shufps` and `addps` pairs. I don't see why it can't use a more efficient instruction selection if it is available. By the way, one other option is a `dpps`-based implementation, which on some processors might actually be slower.

**Comment by mikedn** [comment]
> Well,  Vector.Dot  could use the trivial implementation with  mulps  and  addps  instead of emitting  shufps  and  addps  pairs

Maybe you mean `haddps` instead of `addps`? It's impossible to do a single dot product with only `mulps` and `addps`.

> By the way, one other option is a  dpps -based implementation, which on some processors might actually be slower.

On my Haswell machine `vdpps` is used. It's not clear to me which variant is actually faster. There's very high variance in the resulting timings, I get anything from your result to the opposite result that shows VectorDP being faster.

**Comment by Jorenkv** [comment]
My guess would be that they only support SSE2 instructions right now and therefore can't emit a `haddps` instruction. Given that, the double-shufps makes perfect sense to compute the dot product of a 4-vector. This is the relevant code listing: https://github.com/dotnet/coreclr/blob/efb6696bc3968402d49172cd245c695ae8e24157/src/jit/simdcodegenxarch.cpp#LC1277

As for your manual approach, probably that has better cache latency and/or instruction parallelism? I mean you're not doing equivalently the same thing so I don't think your benchmark is very insightful.

**Comment by goldshtn** [comment]
@Jorenkv @mikedn  At the very least, I hope you agree there shouldn't be any array bounds checks in the above code. The JIT has already a strategy where it generates two versions of the loop -- one with and one without bounds checks. Not sure why this isn't happening here, but that can't be good for perf either.

**Comment by Jorenkv** [comment]
Indeed. The JIT is able to elide the bounds checks for `for (int i = 0; i < a.Length; i++)` and it's obvious to any human that changing `i++` to `i += any positive number` doesn't change anything at all about the validity of `a[i]`: as long as the loop is guarded by `i < a.Length`, `i` started at 0, and is only ever increased then `a[i]` will always be valid. The JIT knows the value of `Vector<float>.Count` at compile time (for a Release build), so there's really no reason why the JIT couldn't elide the bounds checks if it were just a bit smarter about it.

**Comment by Jorenkv** [comment]
So I did some digging into the range checking thing, and found this comment: https://github.com/dotnet/coreclr/blob/4cf8a6b082d9bb1789facd996d8265d3908757b2/src/jit/optimizer.cpp#L7643

    /* We require the loop to add by exactly one.
       The reason for this requirement is that we don't have any guarantees 
       on the length of the array.  For example, if the array length is 0x7fffffff 
       and the increment is by 2, then it's possible for the loop test to overflow
       to 0x80000000 and hence succeed.  In that case it's not legal for us to
       remove the bounds check, because it is expected to fail.
       It is possible to relax this constraint to allow for iterators <= i if we
       can guarantee that arrays will have no more than 0x7fffffff-i+1 elements.
       This requires corresponding logic in the EE that can get out of date with
       the codes here.  We will consider implementing that if this constraint
       causes performance issues. */

I disagree with the premise that relaxing this constraint would require an update to the execution engine to ensure an upper bound on array lengths. If the JIT detects that the stride is unequal to 1 and is therefore worried about the loop iterator overflowing, it could insert code to check this:

```csharp
if ((long)array.Length + n >= int.MaxValue)
{
    // do a bounds checked version of the loop
}
else
{
  // do the loop without bounds checking
}
```

**Comment by goldshtn** [comment]
> If the JIT detects that the stride is unequal to 1 and is therefore worried about the loop iterator overflowing, it could generate code to check this:

@Jorenkv This is especially true since the JIT already has the logic for generating two versions of the same loop and dispatching at runtime based on array lengths. I don't have an example in front of me right now, but off the top of my head the following code would have two runtime paths -- one for when `b.Length <= a.Length` and another when it's larger.

```C#
for (int i = 0; i < a.Length; ++i)
  a[i] = b[i];
```

**Comment by goldshtn** [comment]
@Jorenkv  BTW as for the performance difference: the C++ code generated from `_mm_dp_ps` was noticeably faster than the C# code generated from `Vector.Dot` last time I checked. Source and assembly [in this gist](https://gist.github.com/goldshtn/a1cf8a1d3007bf731224).

**Comment by mikedn** [comment]
>  At the very least, I hope you agree there shouldn't be any array bounds checks in the above code. 

Well, yes but wasn't the main subject of your issue, it was only a side note.

>  it's obvious to any human that changing  i++  to  i += any positive number  doesn't change anything at all about the validity of  a[i] 

It's a bit more complicated than that because in this case there's not only `a[i]`, there's also `a[i+1]`, `a[i+2]` ... `a[i+7]`.

> If the JIT detects that the stride is unequal to 1 and is therefore worried about the loop iterator overflowing, it could insert code to check this:

The question is if doing that is beneficial given the increase in code size required to handle very rare cases.

**Comment by goldshtn** [comment]
@mikedn Take a look at [my comment above](https://github.com/dotnet/coreclr/issues/1414#issuecomment-132637166). I think perf can be affected by the extraneous range-checking, yep.

**Comment by Jorenkv** [comment]
@mikedn: That's true, but from what I can tell from reading the source the JIT already has extensive logic to find the ranges of loop variables and array indices, so it's not a problem it doesn't already deal with. It also already has code to check if a loop variable can overflow. Given this I wouldn't think it's a massive feature request.

It would already be really nice cool that if I wrote `if (a.Length > 9000) return;` (or something else that makes the loop unreachable), the JIT could then reason about the array index bounds sufficiently to elide the bounds checks for this scenario. I'm not sure if it already does anything to this extent, and have no idea how much trouble it would be to add. I would certainly consider it a valuable alternative to the 'proper' analysis.

**Comment by mikedn** [comment]
@goldshtn
> Take a look at my comment above. I think perf can be affected by the extraneous range-checking, yep.

Of course it affects the perf, the question was how much. I tried to compare `VectorDP` with a C++ version that also uses `shufps` and the difference is 2-3%. And if `dpps` is used in the C++ version then the  difference is around 10%.

@Jorenkv
> Given this I wouldn't think it's a massive feature request.

As stated above, my comment wasn't about how easy is this to implement but about the fact that those bound checks may not be significant (in this particular case).

**Comment by CarolEidt** [comment]
@goldshtn  - Sorry for not chiming in sooner. A few observations:
- As @Jorenkv has pointed out, the JIT/EE currently supports only SSE2 or AVX2. This is primarily to reduce the complexity and test matrix.
- I would like to see the bounds checks eliminated in this kind of example (at least through versioning), and I'm hoping that we can make some improvements both in induction variable analysis and range checking. One thing, though, is that (as @mikedn points out) they couldn't be eliminated with the current code because the bounds check would allow Vector<float>.Count - 1 elements to be read beyond the end of the array, if the array were not an even multiple of Vector<float>.Count.

**Comment by goldshtn** [comment]
> Of course it affects the perf, the question was how much. I tried to compare VectorDP with a C++ version that also uses shufps and the difference is 2-3%. And if dpps is used in the C++ version then the difference is around 10%.

@mikedn: We're probably testing on wildly different processors. I compared a C++ implementation with `dpps` to `Vector.Dot` on a DS-series Azure VM, and got a 70% performance difference.

**Comment by redknightlois** [comment]
@goldshtn Not necessarily a different machine, most benchmark code out there suffers from jitter if not properly implemented; specially in cases where the benchmark code is a couple of assembler instructions. For that kind of benchmarks I suggest you to use https://github.com/PerfDotNet/BenchmarkDotNet 

Specifically see this sample (which is aimed to measure single operations): https://github.com/PerfDotNet/BenchmarkDotNet/blob/master/BenchmarkDotNet.Samples/Cpu_Ilp_Inc.cs
